source/common/router/upstream_request.cc

```diff
@@ -114,12 +114,25 @@ void UpstreamRequest::decode100ContinueHeaders(Http::ResponseHeaderMapPtr&& head
   ScopeTrackerScopeState scope(&parent_.callbacks()->scope(), parent_.callbacks()->dispatcher());
 
   ASSERT(100 == Http::Utility::getResponseStatus(*headers));
-  parent_.onUpstream100ContinueHeaders(std::move(headers), *this);
+  // We coalesce 100-continue headers here, to prevent HCM from having to worry about this.
+  if (!decoded_100_continue_headers_) {
+    decoded_100_continue_headers_ = true;
+    parent_.onUpstream100ContinueHeaders(std::move(headers), *this);
+  }
 }
 
 void UpstreamRequest::decodeHeaders(Http::ResponseHeaderMapPtr&& headers, bool end_stream) {
```

Comment:
Also the more I think about this, the more I wonder if this is the right place.

See comments in https://github.com/envoyproxy/envoy/issues/12131

Right now, if an upstream times out without response headers, we can retry.
If an upstream times out after sending 100-continue, we can't.
It'd be kind of nice if we did the 100-continue coalescing somewhere above the upstream request (which is per-upstream-stream), either in the router or the HCM, such that we could actually do retries any time we hadn't sent non-informational headers up.  WDYT?

Reply:
Yeah, fair enough, I'll move this.